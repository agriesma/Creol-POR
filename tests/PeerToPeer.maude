load interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Node" : Cl | Inh: noInh, Par: "db", Att: "db" |-> null, Mtds: 
  < "enquire" : Mtdname | Param: noAid, Latt: "files" |-> null, Code: "sync.0" ! "db" . "listFiles" ( emp ) ; ( await ( "sync.0" ?? ) ) ; ( "sync.0" ? ( "files" ) )  ; return ( "files" ) > *
  < "getLength" : Mtdname | Param: "fId", Latt: "fId" |-> null , "lth" |-> null, Code: "sync.0" ! "db" . "length" ( "fId" ) ; ( await ( "sync.0" ?? ) ) ; ( "sync.0" ? ( "lth" ) )  ; return ( "lth" ) > *
  < "getPack" : Mtdname | Param: "fId" , "pNbr", Latt: "fId" |-> null , "pNbr" |-> null , "pack" |-> null , "f" |-> null, Code: "sync.0" ! "db" . "getFile" ( "fId" ) ; ( await ( "sync.0" ?? ) ) ; ( "sync.0" ? ( "f" ) )  ; "pack" ::= "f" ( "pNbr" ) ; return ( "pack" ) > *
  < "availFiles" : Mtdname | Param: "sList", Latt: "sList" |-> null , "files" |-> null, Code: if "=" ( "sList" # list(emp) ) th "files" ::= list(emp) el "l1" ! "hd" ( "sList" ) . "enquire" ( emp ) ; "l2" ! "self" . "availFiles" ( "tail" ( "sList" ) ) ; ( await "&&" ( "l1" # "l2" ) ) ; ( "l1" ? ( "fList" ) )  ; ( "l2" ? ( "files" ) )  fi ; return ( "files" ) > *
  < "reqFile" : Mtdname | Param: "sId" , "fId", Latt: "sId" |-> null , "fId" |-> null , "file" |-> null , "pack" |-> null , "lth" |-> null, Code: "sync.0" ! "sId" . "getLength" ( "fId" # "lth" ) ; ( await ( "sync.0" ?? ) ) ; ( "sync.0" ? ( "pack" ) )  ; ".anon" ! "db" . "storeFile" ( "fId" # "file" ) ; free( ".anon" ) ; return ( emp ) >, Ocnt: 0 >

.
endm
