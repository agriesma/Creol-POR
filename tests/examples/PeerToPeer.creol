/*
 * 
 */
interface Client
begin
  with Any
    op availFiles (in sList: List[Server];
		   out files: List[[Server , List[Str]]])
    op reqFiles
end

interface Server
begin
  with Server
    op enquire
    op getLength(in fId: String; out lth: Int)
    op getPack(in fId: String, pNbr: Int; out pack: Data)
end

interface Peer
  inherits Client, Server
begin
end

interface DB
begin
  with Server
    op getFile(in fId: String; out file: List[List[Data]])
    op getLength(in fId: String; out length: Int)
    op storeFile(in fId: String, file: List[Data])
    op listFiles(out fList: List[String])
end

class Node(db: DB)
  contracts Peer, Server
begin
  with Server
    op enquire(out files: List[String]) ==
      await db.listFiles(; files)
    op getLength(in fId: String; out lth: Int) ==
      await db.getLength(fId; lth)
    op getPack(in fId: String, pNbr: Int; out pack: Data) ==
      var f: List[List[Data]];
      await db.getFile(fId; f);
      pack := at(f, pNbr)

  with Any
    op availFiles (in sList: List[Server];
		   out files: List[[Server , List[Str]]]) ==
      var l1: Label[ ] ;
      var l2: Label[List[[Server, List[Str]]]] ;
      if (sList = nil) then files := nil
      else
        l1!head(sList).enquire();
        l2!(this as Client).availFiles(tail(sList));
	await l1? && l2?;
	l1?();
	l2?(files)
      end

    op reqFile(in sId: Server, fId: String) ==
      var file: List[Data];
      var pack: Data;
      var lth: Int;
      await sId.getLength(fId; lth);
      while (lth > 0) do
	await sId.getPack(fId, lth; pack);
	file := pack -| file;
	lth := lth - 1
      end;
      !db.storeFile(fId, file)
end

