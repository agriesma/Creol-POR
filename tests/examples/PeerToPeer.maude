load creol-interpreter
mod PROGRAM is
pr CREOL-INTERPRETER .
op init : -> Configuration [ctor] .
eq init =
< "Node" : Cl | Inh: noInh, Par: "db", Att: "db" |-> null, Mtds: 
  < "init" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "run" : Mtdname | Param: noVid, Latt: noSubst, Code: skip ; return ( emp ) > *
  < "enquire" : Mtdname | Param: noVid, Latt: "files" |-> null , "label:0" |-> null, Code: "label:0" ! "db" . "listFiles" ( emp ) ; ( await ( "label:0" ?? ) ) ; ( "label:0" ? ( "files" ) )  ; return ( "files" ) > *
  < "getLength" : Mtdname | Param: "fId", Latt: "fId" |-> null , "lth" |-> null , "label:0" |-> null, Code: "label:0" ! "db" . "getLength" ( "fId" ) ; ( await ( "label:0" ?? ) ) ; ( "label:0" ? ( "lth" ) )  ; return ( "lth" ) > *
  < "getPack" : Mtdname | Param: "fId" , "pNbr", Latt: "fId" |-> null , "pNbr" |-> null , "pack" |-> null , "label:0" |-> null , "f" |-> null, Code: "label:0" ! "db" . "getFile" ( "fId" ) ; ( await ( "label:0" ?? ) ) ; ( "label:0" ? ( "f" ) )  ; "pack" ::= "at" ( "f" # "pNbr" ) ; return ( "pack" ) > *
  < "availFiles" : Mtdname | Param: "sList", Latt: "sList" |-> null , "files" |-> null , "l1" |-> null , "l2" |-> null, Code: if "=" ( "sList" # list(emp) ) th "files" ::= list(emp) el "l1" ! "head" ( "sList" ) . "enquire" ( emp ) ; "l2" ! "this" . "availFiles" ( "tail" ( "sList" ) ) ; ( await "&&" ( ( "l1" ?? ) # ( "l2" ?? ) ) ) ; ( "l1" ? ( noVid ) )  ; ( "l2" ? ( "files" ) )  fi ; return ( "files" ) > *
  < "reqFile" : Mtdname | Param: "sId" , "fId", Latt: "sId" |-> null , "fId" |-> null , "label:2" |-> null , "label:1" |-> null , "label:0" |-> null , "file" |-> null , "pack" |-> null , "lth" |-> null, Code: "label:0" ! "sId" . "getLength" ( "fId" ) ; ( await ( "label:0" ?? ) ) ; ( "label:0" ? ( "lth" ) )  ; while ">" ( "lth" # int(0) ) do "label:1" ! "sId" . "getPack" ( "fId" # "lth" ) ; ( await ( "label:1" ?? ) ) ; ( "label:1" ? ( "pack" ) )  ; "file" ::= "-|" ( "pack" # "file" ) ; "lth" ::= "-" ( "lth" # int(1) ) od  ; "label:2" ! "db" . "storeFile" ( "fId" # "file" ) ; free( "label:2" ) ; return ( emp ) >, Ocnt: 0 >

.
endm
