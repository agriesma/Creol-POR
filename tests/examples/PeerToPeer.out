==========================================
reduce in PROGRAM : classes .
result Class: 
< "Node" : Cl | Inh: noInh, Par: "db", Att: "db" |-> null, Mtds:
  < "availFiles" : Mtdname | Param: "sList", Latt: "_" |-> null, "fList" |->
    null, "files" |-> null, "l1" |-> null, "l2" |-> null, "sList" |-> null,
    Code: if "="("sList" :: list(emp)) th "files" ::= list(emp) el "l1" !
    "head"("sList") . "enquire"(emp) ; "l2" ! "this" . "availFiles"("tail"(
    "sList")) ; await "&&"("l1" ?? :: "l2" ??) ; "l1" ?("fList") ; "l2" ?(
    "files") ; "files" ::= "|-"("files" :: pair("head"("sList"), "fList")) fi ;
    return("files") > *
  < "enquire" : Mtdname | Param: noVid, Latt: "_" |-> null, "files" |-> null,
    "label:0" |-> null, Code: "label:0" ! "db" . "listFiles"(emp) ; await (
    "label:0" ??) ; "label:0" ?("files") ; return("files") > *
  < "getLength" : Mtdname | Param: "fId", Latt: "_" |-> null, "fId" |-> null,
    "label:0" |-> null, "lth" |-> null, Code: "label:0" ! "db" . "getLength"(
    "fId") ; await ("label:0" ??) ; "label:0" ?("lth") ; return("lth") > *
  < "getPack" : Mtdname | Param: "fId", "pNbr", Latt: "_" |-> null, "f" |->
    null, "fId" |-> null, "label:0" |-> null, "pNbr" |-> null, "pack" |-> null,
    Code: "label:0" ! "db" . "getFile"("fId") ; await ("label:0" ??) ;
    "label:0" ?("f") ; "pack" ::= "nth"("f" :: "pNbr") ; return("pack") > *
  < "init" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) > *
  < "reqFile" : Mtdname | Param: "sId", "fId", Latt: "_" |-> null, "fId" |->
    null, "file" |-> null, "label:0" |-> null, "label:1" |-> null, "label:2"
    |-> null, "lth" |-> null, "pack" |-> null, "sId" |-> null, Code: "label:0"
    ! "sId" . "getLength"("fId") ; await ("label:0" ??) ; "label:0" ?("lth") ;
    while ">"("lth" :: int(0)) do "label:1" ! "sId" . "getPack"("fId" :: "lth")
    ; await ("label:1" ??) ; "label:1" ?("pack") ; "file" ::= "-|"("pack" ::
    "file") ; "lth" ::= "-"("lth" :: int(1)) od ; "label:2" ! "db" .
    "storeFile"("fId" :: "file") ; free("label:2") ; return(emp) > *
  < "run" : Mtdname | Param: noVid, Latt: "_" |-> null, Code: skip ; return(
    emp) >,
  Ocnt: 0 >

Bye.
